pkg/storage/cacher.go

```diff
@@ -531,17 +531,23 @@ func (c *Cacher) dispatchEvents() {
 func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	triggerValues, supported := c.triggerValues(event)
 
+	// TODO: For now we assume we have a given <timeout> budget for dispatching
+	// a single event. We should consider changing to the approach with:
+	// - budget has upper bound at <max_timeout>
+	// - we add <portion> to current timeout every second
+	timeout := time.Duration(250) * time.Millisecond
```

Comment:
It's not a full fix, but it would solve the main underlying problem (there won't be hanging for few minutes List requests - because we will be releasing lock at least once every ~250ms). I verified that indeed that one is fixed and extended some test. It doesn't protect for really stale data in watch, but I think it's slightly too late for such fix now. I will try to do it next week.